plugins.d/Lets_Encrypt/dns_01.py

```diff
@@ -3,73 +3,197 @@
+                   " removing deb and install via venv.")
+    else:
+        msg_mid = "but your system is in an unexpected state"
+    if msg is not None:
+        msg = msg_start + msg_mid + msg_end
+        ret = console.yesno(msg, autosize=True)
         if ret != 'ok':
-            return None, 'Please install lexicon to use dns-01 challenge.'
-
-        apt = subprocess.run(['apt-get', '-y', 'install', 'lexicon'],
-                             encoding=sys.stdin.encoding,
-                             stderr=PIPE)
-        if apt.returncode != 0:
-            return None, apt.stderr.strip()
-
+            return
+    if remove_lexicon:
+        print("Please wait while lexicon package is removed")
+        exit_code, msg = run_command(['apt-get', 'remove', '-y', 'lexicon'])
+        if exit_code != 0:
+            console.msgbox(
+                'Error',
+                f"Removal of lexicon package failed:\n\n{msg}")
+    if install_venv:
+        pkgs = []
+        pip = which('pip')
+        python3_venv = check_pkg('python3-venv')
+        if not pip:
+            pkgs.append('python3-pip')
+        if not python3_venv:
+            pkgs.append('python3-venv')
+        if pkgs:
+            print("Please wait while required packages are installed")
+            exit_code, msg = apt_install(pkgs)
+            if exit_code != 0:
+                pkgs_l = " ".join(pkgs)
+                console.msgbox(
+                    'Error',
+                    f"Apt installing {pkgs_l} failed:\n\n{msg}")
+        makedirs(dirname(venv), exist_ok=True)
+        local_bin = '/usr/local/bin'
+        venv_pip = join(venv, 'bin/pip')
+        for comment_command in [
+                ("venv is set up", ['/usr/bin/python3', '-m', 'venv', venv]),
+                ("lexicon is installed (into venv)", [venv_pip, 'install', 'dns-lexicon[full]']),
+                (None, ['/usr/bin/ln', '-sf', f"{venv}/bin/lexicon", f"{local_bin}/lexicon"]),
+                (None, ['/usr/bin/ln', '-sf', f"{venv}/bin/tldextract", f"{local_bin}/tldextract"])
+            ]:
+            comment, command = comment_command
+            assert isinstance(command, list)
+            if comment:
+                print(f"Please wait while {comment}")
+            exit_code, msg = run_command(command)
+            if exit_code != 0:
+                com = " ".join(command)
+                console.msgbox(
+                    'Error',
+                    f"Command '{com}' failed:\n\n{msg}")
+                return None
+
+        lexicon_bin = which('lexicon')
+        lexicon_loc = '/usr/local/bin/lexicon'
+        if lexicon_bin:
+            if lexicon_bin != lexicon_loc:
+                console.msgbox(
+                    'Error',
+                    f"lexicon ({lexicon_bin}) not found where expected ({lexicon_loc})?!")
+            else:
+                return None
+        console.msgbox('Error', "Something went wrong... Please report to TurnKey.")
```

Comment:
Can we just avoid dealing with a system installed lexicon entirely?

Reply:
Yep, I followed your suggestion.